### PR TITLE
Integrate Keystone - Fix issues, add events tracking

### PR DESCRIPTION
**Pull request by soralit**
**Description**
This PR resolves the most issues we found in the previous round testing, and also add events tracking.

### CODE REVIEW

File: app/components/UI/AccountList/AccountElement/index.js [hunk 1]
```diff
@@ -120,14 +120,20 @@ class AccountElement extends PureComponent {
 
 	render() {
 		const { disabled, updatedBalanceFromStore, ticker } = this.props;
-		const { address, name, ens, isSelected, isImported, balanceError } = this.props.item;
+		const { address, name, ens, isSelected, isImported, balanceError, isQRHardware } = this.props.item;
 		const selected = isSelected ? <Icon name="check-circle" size={30} color={colors.blue} /> : null;
-		const imported = isImported ? (
+		const tag = isImported ? (
```
Comment (andreahaku) [inline app/components/UI/AccountList/AccountElement/index.js:125]: what about:

```
const tag = (isImported || isQRHardware) ? (
    <View style={styles.importedWrapper}>
        <Text numberOfLines={1} style={styles.importedText}>
            {isImported ? strings('accounts.imported') : strings('transaction.hardware')}
        </Text>
    </View>
) :  null;
```

File: app/components/UI/AccountList/AccountElement/index.js [hunk 2]
```diff
@@ -156,7 +162,7 @@ class AccountElement extends PureComponent {
 								)}
 							</View>
 						</View>
-						{!!imported && <View style={styles.importedView}>{imported}</View>}
+						{!!tag && <View style={styles.importedView}>{tag}</View>}
```
Comment (andreahaku) [inline app/components/UI/AccountList/AccountElement/index.js:165]: why you're wrapping `tag` with another `<View style={styles.importedView}>`?
It should simply be:
`{tag}`

File: app/components/UI/AccountOverview/index.js [hunk 3]
```diff
@@ -349,21 +371,28 @@ class AccountOverview extends PureComponent {
...
-										{isDefaultAccountName(name) && ens ? ens : name}
-									</Text>
-								</TouchableOpacity>
+								<View style={styles.labelWrapper}>
+									<TouchableOpacity onLongPress={this.setAccountLabelEditable}>
+										<Text
+											style={[
+												styles.label,
+												styles.onboardingWizardLabel,
+												onboardingWizard
```
Comment (andreahaku) [inline app/components/UI/AccountOverview/index.js:380]: suggested:
`{ borderColor: onboardingWizard ? colors.blue : colors.white}`

File: app/components/UI/AccountOverview/index.js [hunk 4]
```diff
@@ -349,21 +371,28 @@ class AccountOverview extends PureComponent {
...
+													? { borderColor: colors.blue }
+													: { borderColor: colors.white },
+											]}
+											numberOfLines={1}
+											testID={'edit-account-label'}
+										>
+											{isDefaultAccountName(name) && ens ? ens : name}
```
Comment (andreahaku) [inline app/components/UI/AccountOverview/index.js:387]: suggested:
`{isDefaultAccountName(name) && (ens || name)}`

File: app/components/UI/QRHardware/AnimatedQRScanner.tsx [hunk 5]
```diff
@@ -10,6 +10,8 @@ import Icon from 'react-native-vector-icons/Ionicons';
 import { strings } from '../../../../locales/i18n';
 import { URRegistryDecoder } from '@keystonehq/ur-decoder';
 import Modal from 'react-native-modal';
+import { UR } from '@ngraveio/bc-ur';
```
Comment (andreahaku) [inline app/components/UI/QRHardware/AnimatedQRScanner.tsx:13]: this import is only needed for the type?

File: app/components/UI/QRHardware/QRSigningDetails.tsx [hunk 6]
```diff
@@ -8,6 +8,11 @@ import { colors, fontStyles } from '../../../styles/common';
 import AccountInfoCard from '../AccountInfoCard';
 import ActionView from '../ActionView';
 import { IQRState } from './types';
+import { UR } from '@ngraveio/bc-ur';
```
Comment (andreahaku) [inline app/components/UI/QRHardware/QRSigningDetails.tsx:11]: same as above

File: app/components/UI/QRHardware/AnimatedQRScanner.tsx [hunk 7]
```diff
@@ -104,39 +118,82 @@ const AnimatedQRScannerModal = (props: AnimatedQRScannerProps) => {
...
+				urDecoder.receivePart(content);
+				setProgress(Math.ceil(urDecoder.getProgress() * 100));
+				if (urDecoder.isError()) {
+					AnalyticsV2.trackEvent(AnalyticsV2.ANALYTICS_EVENTS.HARDWARE_WALLET_ERROR, {
+						...getAnalyticParams(),
+						error: urDecoder.resultError(),
+					});
+					onScanError(strings('transaction.unknown_qr_code'));
 				}
+				if (urDecoder.isSuccess()) {
```
Comment (andreahaku) [inline app/components/UI/QRHardware/AnimatedQRScanner.tsx:157]: else if

File: app/components/UI/QRHardware/QRSigningDetails.tsx [hunk 8]
```diff
@@ -77,40 +100,69 @@ const QRSigningDetails = ({
...
+	showHint = true,
 }: IQRSigningDetails) => {
 	const KeyringController = useMemo(() => {
 		const { KeyringController: keyring } = Engine.context as any;
 		return keyring;
 	}, []);
 	const [scannerVisible, setScannerVisible] = useState(false);
+	const [errorMessage, setErrorMessage] = useState('');
+
+	const resetError = useCallback(() => {
```
Comment (andreahaku) [inline app/components/UI/QRHardware/QRSigningDetails.tsx:112]: probably the `useCallback` is not needed here

File: app/components/UI/QRHardware/QRSigningDetails.tsx [hunk 9]
```diff
@@ -77,40 +100,69 @@ const QRSigningDetails = ({
...
+		setErrorMessage('');
+	}, []);
+
 	const showScanner = useCallback(() => {
 		setScannerVisible(true);
-	}, []);
+		resetError();
+	}, [resetError]);
+
 	const hideScanner = useCallback(() => {
```
Comment (andreahaku) [inline app/components/UI/QRHardware/QRSigningDetails.tsx:121]: same here

File: app/components/Views/Approval/index.js [hunk 10]
```diff
@@ -258,10 +263,14 @@ class Approval extends PureComponent {
 			await TransactionController.approveTransaction(transaction.id);
 			this.showWalletConnectNotification(true);
 		} catch (error) {
-			Alert.alert(strings('transactions.transaction_error'), error && error.message, [
-				{ text: strings('navigation.ok') },
-			]);
-			Logger.error(error, 'error while trying to send transaction (Approval)');
+			if (!error.message.startsWith('KeystoneError#Tx_canceled')) {
+				Alert.alert(strings('transactions.transaction_error'), error && error.message, [
```
Comment (andreahaku) [inline app/components/Views/Approval/index.js:267]: suggested:
`error?.message`

File: app/components/Views/ApproveView/Approve/index.js [hunk 11]
```diff
@@ -442,8 +442,12 @@ class Approve extends PureComponent {
 			await TransactionController.approveTransaction(transaction.id);
 			AnalyticsV2.trackEvent(AnalyticsV2.ANALYTICS_EVENTS.APPROVAL_COMPLETED, this.getAnalyticsParams());
 		} catch (error) {
-			Alert.alert(strings('transactions.transaction_error'), error && error.message, [{ text: 'OK' }]);
-			Logger.error(error, 'error while trying to send transaction (Approve)');
+			if (!error.message.startsWith('KeystoneError#Tx_canceled')) {
+				Alert.alert(strings('transactions.transaction_error'), error && error.message, [{ text: 'OK' }]);
```
Comment (andreahaku) [inline app/components/Views/ApproveView/Approve/index.js:446]: suggested:
`error?.message`

**Review by andreahaku — COMMENTED** [review]
mainly only a few minor suggested code improvements

**Review by gantunesr — COMMENTED** [review]
This is looking good Sora! @andreahaku left some observations and I added some more. Other points to have in mind:

1. We have to bring the analytics events to QA (@plasmacorral) so its included in their test
2. Looks like the `yarn audit` stage is failing. Let me know if you need help with it

File: app/util/networks/index.js [hunk 12]
```diff
@@ -103,6 +104,18 @@ export function getNetworkTypeById(id) {
...
+	if (!chainId) {
+		throw new Error(NETWORK_ERROR_MISSING_CHAIN_ID);
+	}
+	const network = NetworkListKeys.filter((key) => NetworkList[key].chainId === parseInt(chainId, 10));
+	if (network.length > 0) {
+		return network[0];
+	}
+
+	throw new Error(`${NETWORK_ERROR_UNKNOWN_CHAIN_ID} ${chainId}`);
+}
```
[the remark this answers is not in the document]
Comment (soralit) [reply app/util/networks/index.js:117]: Cool

File: app/components/UI/AccountList/index.js [hunk 13]
```diff
@@ -263,7 +264,19 @@ class AccountList extends PureComponent {
 		let ret = false;
 		for (const keyring of allKeyrings) {
 			if (keyring.accounts.includes(address)) {
-				ret = keyring.type !== 'HD Key Tree';
+				ret = keyring.type === KeyringTypes.simple;
```
[the remark this answers is not in the document]
Comment (soralit) [reply app/components/UI/AccountList/index.js:267]: Just modified existing code, maybe `ret` = `return`.

File: app/components/Nav/Main/index.js [hunk 14]
```diff
@@ -299,10 +300,14 @@ const Main = (props) => {
 				});
 				await TransactionController.approveTransaction(transactionMeta.id);
 			} catch (error) {
-				Alert.alert(strings('transactions.transaction_error'), error && error.message, [
-					{ text: strings('navigation.ok') },
-				]);
-				Logger.error(error, 'error while trying to send transaction (Main)');
+				if (!error.message.startsWith('KeystoneError#Tx_canceled')) {
```
[the remark this answers is not in the document]
Comment (soralit) [reply app/components/Nav/Main/index.js:303]: Currently no, the only error would happen in the integration is `Tx_canceled`, which identify that user canceled the signing process and inform UI to do changes.

File: app/components/Views/ApproveView/Approve/index.js (hunk 11 above)
Comment (soralit) [reply app/components/Views/ApproveView/Approve/index.js:446]: Thanks.

File: app/components/UI/QRHardware/AnimatedQRScanner.tsx (hunk 5 above)
Comment (soralit) [reply app/components/UI/QRHardware/AnimatedQRScanner.tsx:13]: Yes.

File: app/components/UI/AccountList/AccountElement/index.js (hunk 2 above)
Comment (soralit) [reply app/components/UI/AccountList/AccountElement/index.js:165]: Hey @andreahaku , the `<View>` inside `tag` has style "importedWrapper" but not "importedView"

File: app/components/UI/AccountList/AccountElement/index.js (hunk 1 above)
Comment (soralit) [reply app/components/UI/AccountList/AccountElement/index.js:125]: Thanks.

**Comment by soralit** [comment]
Hey @gantunesr , seems the `yarn audits` fails for security issues about the installed packages, I see some vulnerabilities were just updated several hours ago. 
Maybe we can create another PR to resolve the audit issues, and rebase the code in this PR. 
Can you offer a help to do that?

**Comment by gantunesr** [comment]
@soralit your dependencies are outdated compared to our main branch, updating it should solve the issues

cc @rickycodes

**Comment by rickycodes** [comment]
@soralit I actually have a rebased copy locally from where you left off @ `c842fed` that's up to date and passing. I just had to update the lockfile. let me know if you want me to push it up here :construction_worker:

**Comment by soralit** [comment]
Hi @rickycodes , can you push the rebased code here? Maybe it is impossible because the branch is in our repo, or you can update the base branch `feat/keystone-integration` so I can update this PR. I will undo the changes and

**Comment by gantunesr** [comment]
Hey @soralit, I didn't realized that `feat/keystone-integration` was outdated, just pushed the changes so you can update your branch

**Review by andreahaku — APPROVED** [review]
Thanks for all the updates. Good work! ;)

**Pull request merged**